core/src/main/java/org/apache/iceberg/RemoveSnapshots.java

```diff
@@ -195,6 +299,18 @@ public void commit() {
     LOG.info("Committed snapshot changes");
 
     if (cleanExpiredFiles) {
+      TableMetadata updated = ops.refresh();
+      if (updated.refs() != null) {
+        List<SnapshotRef> branches = updated.refs()
+            .values().stream().filter(SnapshotRef::isBranch)
+            .collect(Collectors.toList());
+
+        if (branches.size() > 1) {
+          throw new UnsupportedOperationException(
+              "Deleting expired files when there is more than 1 branch is currently not supported");
+        }
```

Comment:
Should this throw after doing all of the work (e.g. the commit etc)?

If I understand correctly, throwing here will cause the actual remove snapshots query / action to "fail" from the POV of the end user.

If that's the case, i think we should log an error and then proceed given that we've already committed the snapshot changes, or check this earlier on and then throw if need be.

Reply:
The problem is that we have several cases that can't be reliably cleaned up without using the reference set. For example:

```
  A - B - C
  ` - D
```

If A is aged off, there is no way to know that B, C, and D share all of the files from A. So expiring B could delete files that are referenced by D but not C.

To take care of cases like this, we need to do reference set comparison. But rather than implementing that all at once here, we decided to fail snapshot expiration in the short term.

However, one thing does need to be fixed: this check should happen _before_ committing the changes. If we detect that we can't reliably clean up the table, we shouldn't drop the snapshots anyway. So this check should come first in the commit. (@amogh-jahagirdar)

Reply:
@rdblue Thanks for the explanation on why the reachability analysis needs to be done and why we can't delete files if there are multiple branches.

> However, one thing does need to be fixed: this check should happen before committing the changes. If we detect that we can't reliably clean up the table, we shouldn't drop the snapshots anyway. So this check should come first in the commit. 

I was thinking about this before, I'm not sure if the operation to remove snapshots should be coupled with the ability to clean up the files. We know reliably which snapshots which should be removed based on the updated algorithm so it seems a bit unnecessarily constraining to fail the entire operation in case we can't do the deletion of files.

Currently, if I'm not mistaken, deletion of the files themselves is already best effort and can fail, so it's already not a guarantee that we can reliably clean up the table. So following this, I think it makes sense to allow the removal of snapshot metadata and then prevent the deletion of files in the case of multiple branches. Right now we're preventing the operation by throwing an exception which as @kbendick pointed out, to a user would seem like the operation failed but it didn't from a snapshot removal perspective. Another way to handle this is to ignore the cleanFiles flag in case there are multiple branches and warn log that nothing is being deleted because of the current constraint (instead of throwing an exception). @kbendic @rdblue Let me know your thoughts!

Reply:
I re-read and thought about @rdblue example. The problem is I was assuming there would be another branch and so I was thinking we would always be able to tell when we can't delete the files just based off the number of refs but that we could remove the snapshots themselves.

My assumption there isn't valid; the 2nd branch's ref could be aged off independently and there would then be an isolated snapshot lineage. Then later on when the procedure is run  we have 

`
   <main> B - C
   D
`

There's one ref (let's say the B-C lineage is main). When we go to expire there still could be files that B and C does not refer to but D does. There is no longer a 2nd branch, so the removal of files would go through and delete files which D has. 

So then yes, we must prevent the actual snapshot removal operation from going through in case of multiple refs.